fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/SynchronizedSessionTest.java

```diff
@@ -0,0 +1,325 @@
+            Thread longSession = new Thread(() -> {
+                for (int i = 0; i < 10; i++) {
+                    try {
+                        Thread.sleep(500);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                    SynchronizedSessionRunner runner = reuseRunner ? session1Runner0 : database.newRunner()
+                            .toSynchronized(lockSubspace, session1, 1_000);
+                    checkActive(runner);
+                }
+                session1Stopped.set(true);
+            });
+            Thread tryStartSession = new Thread(() -> {
+                while (!session1Stopped.get()) {
+                    assertFailedStartSession(lockSubspace);
+                    try {
+                        Thread.sleep(random.nextInt(500));
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                }
+            });
+            longSession.start();
+            tryStartSession.start();
+            tryStartSession.join();
+        }
+    }
+
+    private SynchronizedSessionRunner newRunnerStartSession(Subspace lockSubspace) {
+        return database.newRunner().toSynchronized(lockSubspace, DEFAULT_LEASE_LENGTH_MILL);
+    }
+
+    private SynchronizedSessionRunner newRunnerJoinSession(Subspace lockSubspace, UUID sessionId) {
+        return database.newRunner().toSynchronized(lockSubspace, sessionId, DEFAULT_LEASE_LENGTH_MILL);
+    }
+
+    private void checkActive(SynchronizedSessionRunner runner) {
+        if (runAsync) {
+            runner.runAsync(c -> AsyncUtil.DONE);
+        } else {
+            runner.run(c -> null);
+        }
+    }
+
+    private void assertFailedStartSession(Subspace lockSubspace) {
+        SynchronizedSessionExpiredException exception = assertThrows(SynchronizedSessionExpiredException.class,
+                () -> newRunnerStartSession(lockSubspace));
+        assertEquals("Failed to initialize the session", exception.getMessage());
+    }
+
+    private void assertFailedJoinSession(Subspace lockSubspace, UUID sessionId) {
+        // It doesn't fail when the runner is created because there is no checking, but when it runs.
+        try (SynchronizedSessionRunner expiredSessionsRunner = newRunnerJoinSession(lockSubspace, sessionId)) {
+            assertFailedContinueSession(expiredSessionsRunner);
+        }
+    }
+
+    private void assertFailedContinueSession(SynchronizedSessionRunner synchronizedSessionRunner) {
+        SynchronizedSessionExpiredException exception = assertThrows(SynchronizedSessionExpiredException.class,
+                () -> synchronizedSessionRunner.run(c -> null));
+        assertEquals("Failed to continue the session", exception.getMessage());
+    }
+
+    private void waitLongEnough() throws InterruptedException {
+        Thread.sleep(DEFAULT_LEASE_LENGTH_MILL + 100);
```

Comment:
Using sleep here in this test does make me somewhat uneasy, just on a philosophical level. I think the only alternative would be to mock the clock, which maybe is worth doing someday.